### ISSUE TITLE
[HOLD for payment 2022-12-20][$2000] [mWeb] keyboard hides while converting Composer to full size - reported by @gadhiyamanan

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#11093]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. go to any chat
2. type multiple lines
3. click on expand button

## Expected Result:
keyboard should not hide

## Actual Result:
keyboard hides

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** 1.1.99-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/190953705-bee968f2-39e2-42c1-9924-3b0cbe848223.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1661857234117209

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by sakluger** [comment]
@neil-marcellini mentioned in Slack that this has been a known issue for a while, but hasn't been reported before. Probably can be worked on by an external contributor.

**Comment by johnmlee101** [comment]
Yep agreed, will open this up.

**Comment by luacmartins** [comment]
Still waiting on a solution for this issue - https://github.com/Expensify/App/issues/11093#issuecomment-1255432667

**Comment by dhairyasenjaliya** [comment]
@parasharrajat I think we should hold this issue in favor of https://github.com/Expensify/App/issues/8592 as I believe if we solve that blue opacity issue eventually we need the same solution here as the ultimate goal is to hold the keyboard in an open state while sending a message, resizing composer along with removing blue opacity for safari

**Comment by parasharrajat** [comment]
I don't think we need to hold this. Someone might be able to find a different solution.

Thanks for linking that issue, I was trying to find that.

**Comment by dhairyasenjaliya** [comment]
Ah probably it will get wider view to look😅

**Comment by luacmartins** [comment]
Still looking for proposals

**Comment by MitchExpensify** [comment]
Doubled!

**Comment by luacmartins** [comment]
Still looking for proposals!

**Comment by JmillsExpensify** [comment]
Let's bump the price of this issue to $1000 to make sure it gets done.

**Comment by MitchExpensify** [comment]
[Doubled! ](https://expensify.slack.com/archives/C01GTK53T8Q/p1667351910960579)

**Comment by eh2077** [comment]
**Proposal**

Hi there,

I'd like to propose a solution by improving the onBlur handling of Composer. The key idea is to filter the blur event triggered by pressing expand/collapse icon and set focus back to the input. One trick worth mentioning is that we should trigger `.focus()` immediately after the handling of expand/collapse click on iOS platform.

The change of onBlur handling will be like
https://github.com/Expensify/App/blob/c2e4ab25a9b2b7d47004d7b323116f32c181a189/src/pages/home/report/ReportActionCompose.js#L653

```diff
- onBlur={() => this.setIsFocused(false)}
+ onBlur={(event) => {
+    // We want to prevent composer from losing (keyboard)focus after pressing the expand or collapse icon
+    // Save a state and the handling of expand/collapse click will set focus back to the Composer
+    if (_.contains([this.composerSizingCollapseID, this.composerSizingExpandID], lodashGet(event, 'nativeEvent.relatedTarget.id'))) {
+        this.setTextInputShouldFocus(true);
+    }
+    this.setIsFocused(false);
+ }}
```
which is inspired by 
https://github.com/Expensify/App/blob/c2e4ab25a9b2b7d47004d7b323116f32c181a189/src/pages/home/report/ReportActionItemMessageEdit.js#L228-L232

Below are the demo videos
Android:

https://user-images.githubusercontent.com/117511920/201146176-1d284df6-2c11-43e1-9819-f3613f922ec6.mp4

iOS:



https://user-images.githubusercontent.com/117511920/201146439-3064c23b-595f-463a-b09b-5c7baf1c3477.mp4

**Comment by eh2077** [comment]
> @eh2077 Please try to expand/collapse when cursor is in the middle or selected some text. Cursor is at the end on your videos so it's not strong test case

Hi @aimane-chnaif, I just added a video about editing text on iOS for your information 

https://user-images.githubusercontent.com/117511920/201162555-79d9be23-a36f-4d45-98ba-d832195cb758.mp4

**Comment by parasharrajat** [comment]
Thanks for the notes @aimane-chnaif. @eh2077 Please check the old discussion for existing challenges and issues.

**Comment by eh2077** [comment]
@aimane-chnaif Thanks for pointing that out. I'll learn more from the old discussions.

**Comment by MitchExpensify** [comment]
Doubled!

@eh2077 Any estimate on another proposal out of interest?

**Comment by MitchExpensify** [comment]
Bump on the above @eh2077

**Comment by hammad-6304** [comment]
@MitchExpensify have this issue resolved or on hold atm or still accepting new proposals on it?

**Comment by parasharrajat** [comment]
Accepting proposals...

**Comment by MitchExpensify** [comment]
[Doubled!](https://expensify.slack.com/archives/C01GTK53T8Q/p1669067109574519?thread_ts=1661857234.117209&cid=C01GTK53T8Q)

**Comment by parasharrajat** [comment]
Let's hold this issue. I thought I mentioned that earlier. There is a challenge that is preventing us to pick a good solution. We are trying to tackle that at https://github.com/Expensify/App/issues/8592. Please reset the price back to $2K. If we are to fix the issue, then it will be done asap. @MitchExpensify

**Comment by luacmartins** [comment]
Coming back from ooo today, I actually think that @eh2077 solution is good here. @parasharrajat I think the bug with the spellcheck highlight exists whether we fix this issue with the keyboard hiding or not, is that correct?

**Comment by parasharrajat** [comment]
I believe that we can prevent the focus from being lost from the composer which will fix this issue. But it has a spellcheck highlight issue which is trying to fix at https://github.com/Expensify/App/issues/8592. So IMO, we should wait for a few days until, we decide the next steps on https://github.com/Expensify/App/issues/8592.

**Comment by luacmartins** [comment]
>So IMO, we should wait for a few days until, we decide the next steps on https://github.com/Expensify/App/issues/8592.

From [this Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1669139073397899), we decided to move this issue along regardless of the spellcheck highlight issue.

Sorry @aimane-chnaif I missed your proposal in my previous comment. I just tested your proposal and it's simple and solves the issue! Let's move on with it!

**Comment by parasharrajat** [comment]
In that case, I also agree @aimane-chnaif's proposal is best here.

**Comment by MitchExpensify** [comment]
[New Upwork Job created](https://www.upwork.com/jobs/~01fc3048f9ceaa2f2a). Invited everyone for eventual payment

**Comment by parasharrajat** [comment]
Although I have tested it already but will give it a full test run tomorrow.

**Comment by MitchExpensify** [comment]
I reverted the price doubling since we actually accepted an old proposal. Bit of a weird one but that seems fair

**Comment by eh2077** [comment]
Congrats to @aimane-chnaif 

While I'd like to mention a potential UX advantage my proposal might provide. When there's a scrollbar in the composer, by calling `textInput.focus` after the click of expand/collapse, the input cursor/caret will be shown on the screen automatically.

Below is an example to compare the difference of caret display when collapsing the composer 

onMouseDown preventDefault | Call `textInput.focus` after the click of expand/collapse
:-: | :-:
<video src='https://user-images.githubusercontent.com/117511920/203451134-c6f91b23-9ec5-47bd-babb-81e100d0df40.mp4' width=180/> | <video src='https://user-images.githubusercontent.com/117511920/203451120-9b4b1325-f757-467e-b74c-36bad19f96bf.mp4' width=180/>

I'll be happy to hear your thoughts on this cc @luacmartins @parasharrajat

**Comment by parasharrajat** [comment]
Hmm, that is surely a valid point. I will dig into this a bit more tomorrow and let you know if something changes here. 

Meanwhile, I will encourage you to look at other open issues @eh2077.

**Comment by eh2077** [comment]
Hey @aimane-chnaif 

> That difference is only from mSafari. 

I found that the calling `focus` method has differences on Safari(both desktop and mobile) and Firefox(both desktop and mobile) based on my testing. There's no difference from Chrome(both desktop and mobile)

> Each solution has drawback since both are beyond the default behavior of web. 

I can't agree more on this.

> but your solution never works on Firefox (both desktop and mobile) which is critical.

This should be not true based on my testing. 

Firefox mobile
![11093-FF-Android](https://user-images.githubusercontent.com/117511920/203500073-96fd8f09-8ad7-41cf-b11b-953a30c13131.gif)

Firefox desktop

https://user-images.githubusercontent.com/117511920/203501560-0a9489fd-2876-47bc-901a-6aefa31270ac.mov




Safari desktop

https://user-images.githubusercontent.com/117511920/203501461-6c9d1874-6e95-4141-8dc4-8dfaac732728.mov




Below diff will be helpful if you would like to test what I mentioned.

```diff
diff --git a/src/pages/home/report/ReportActionCompose.js b/src/pages/home/report/ReportActionCompose.js
index 2335864f91..0f566031f8 100644
--- a/src/pages/home/report/ReportActionCompose.js
+++ b/src/pages/home/report/ReportActionCompose.js
@@ -128,6 +128,8 @@ class ReportActionCompose extends React.Component {
         this.getInputPlaceholder = this.getInputPlaceholder.bind(this);
         this.getIOUOptions = this.getIOUOptions.bind(this);
         this.addAttachment = this.addAttachment.bind(this);
+        this.composerSizingCollapseID = 'composerSizingCollapseID';
+        this.composerSizingExpandID = 'composerSizingExpandID';
 
         this.comment = props.comment;
         this.shouldFocusInputOnScreenFocus = canFocusInputOnScreenFocus();
@@ -136,6 +138,7 @@ class ReportActionCompose extends React.Component {
             isFocused: this.shouldFocusInputOnScreenFocus,
             isFullComposerAvailable: props.isComposerFullSize,
             textInputShouldClear: false,
+            textInputShouldFocus: false,
             isCommentEmpty: props.comment.length === 0,
             isMenuVisible: false,
             selection: {
@@ -219,6 +222,10 @@ class ReportActionCompose extends React.Component {
         this.setState({textInputShouldClear: shouldClear});
     }
 
+    setTextInputShouldFocus(shouldFocus) {
+        this.setState({textInputShouldFocus: shouldFocus});
+    }
+
     /**
      * Updates the visibility state of the menu
      *
@@ -561,8 +568,13 @@ class ReportActionCompose extends React.Component {
                                                             onPress={(e) => {
                                                                 e.preventDefault();
                                                                 Report.setIsComposerFullSize(this.props.reportID, false);
+                                                                if (this.state.textInputShouldFocus) {
+                                                                    this.textInput.focus();
+                                                                    this.setTextInputShouldFocus(false);
+                                                                }
                                                             }}
                                                             style={styles.composerSizeButton}
+                                                            nativeID={this.composerSizingCollapseID}
                                                             underlayColor={themeColors.componentBG}
                                                             disabled={isBlockedFromConcierge}
                                                         >
@@ -577,8 +589,13 @@ class ReportActionCompose extends React.Component {
                                                             onPress={(e) => {
                                                                 e.preventDefault();
                                                                 Report.setIsComposerFullSize(this.props.reportID, true);
+                                                                if (this.state.textInputShouldFocus) {
+                                                                    this.textInput.focus();
+                                                                    this.setTextInputShouldFocus(false);
+                                                                }
                                                             }}
                                                             style={styles.composerSizeButton}
+                                                            nativeID={this.composerSizingExpandID}
                                                             underlayColor={themeColors.componentBG}
                                                             disabled={isBlockedFromConcierge}
                                                         >
@@ -664,7 +681,14 @@ class ReportActionCompose extends React.Component {
                                         style={[styles.textInputCompose, this.props.isComposerFullSize ? styles.textInputFullCompose : styles.flex4]}
                                         maxLines={this.state.maxLines}
                                         onFocus={() => this.setIsFocused(true)}
-                                        onBlur={() => this.setIsFocused(false)}
+                                        onBlur={(event) => {
+                                            // We want to prevent composer from losing (keyboard)focus after pressing the expand or collapse icon
+                                            // Save a state and the handling of expand/collapse click will set focus back to the Composer
+                                            if (_.contains([this.composerSizingCollapseID, this.composerSizingExpandID], lodashGet(event, 'nativeEvent.relatedTarget.id'))) {
+                                                this.setTextInputShouldFocus(true);
+                                            }
+                                            this.setIsFocused(false);
+                                        }}
                                         onPasteFile={displayFileInModal}
                                         shouldClear={this.state.textInputShouldClear}
                                         onClear={() => this.setTextInputShouldClear(false)}
```

**Comment by eh2077** [comment]
@aimane-chnaif Thanks for the testing! Your last comment is basically fair while it could be minor too.

**Comment by luacmartins** [comment]
PR is merged. Just waiting for the deploys!

**Comment by MitchExpensify** [comment]
I'm heading ooo so reassigning the Bug label for a new CM!

**Comment by luacmartins** [comment]
@zanyrenney I'm not sure why the title wasn't updated, but this PR was [deployed to prod on Dec 13](https://github.com/Expensify/App/pull/12949#issuecomment-1349933101) so the regression period ends today!

**Comment by zanyrenney** [comment]
awesome, paying this one out and closing.

**Comment by zanyrenney** [comment]
Hey @aimane-chnaif thanks for that, I had the upwork link handy, we're just having some internal discussions. FYI for Expensifiers: [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1671642340300909)

**Comment by zanyrenney** [comment]
I'm heading OOO and we're still discussing this here, so reapplying the Bug0 label in this case.

**Comment by puneetlath** [comment]
All paid!
